libdnf/transaction/transaction_item_reason.cpp

```diff
@@ -65,15 +65,12 @@ TransactionItemReason transaction_item_reason_from_string(const std::string & re
     throw InvalidTransactionItemReason(reason);
 }
 
-// TransactionItemReason::UNKNOWN will have higher value than TransactionItemReason::DEPENDENCY
-// Important for autoremove to prevent unwanted removal
-// https://bugzilla.redhat.com/show_bug.cgi?id=1921063
-// TODO(lukash) the above doesn't make much sense. likely not needed with the new system state, revise
+
 static TransactionItemReason order[] = {
+    TransactionItemReason::NONE,
```

Comment:
This part is used for goal to determine the proper reason with obsoletes. I think that the original position of none is correct.
```
        for (int i = 0; i < obs.size(); ++i) {
            rpm::Package obsoleted(base, rpm::PackageId(obs[i]));
            auto obs_reson = obsoleted.get_reason();
            if (obs_reson > reason) {
                reason = obs_reson;
            }
```

Reply:
It's not correct, it's a horrible hack, unfortunately. If you know of a case in which this will fail, please describe it so that I can create a test scenario for it.

As for a solution, I consider special-casing `None` in a particular case better than changing the otherwise logical order of the reasons. Note that now the `get_reason()` method returns `None` when there's no record for the NA in `packages.toml`. That can mean two things though, either the package is not installed, or it was installed outside dnf and the "effective" reason is `User`. I've even been considering adding a `get_effective_reason()` method to `Package`, but using that in the `base::Transaction` logic wouldn't work well AFAICS as that still erases the above-described distinction.

Reply:
Well thank you for your description because I realized that the issue is a different. The issue is that None reason represents two reasons and we have to distinguish between them. Please can you add a reason representing installed outside of DNF stack?

Reply:
The main user of reason comparison is Transaction class where only usage of None is for packages installed outside of DNF because the comparison only take place where installed package is replaced.